### PR TITLE
feat(train): 細かい機能追加

### PR DESCRIPTION
**Pull request by sevenc-nanashi**
## 内容

- lrをconfigで指定できるようにします。
- Loss/evalをTensorboardにプロットするようにします。

## 関連 Issue

（なし）

## スクリーンショット・動画など

（なし）

## その他

（なし）

### CODE REVIEW

File: train/src/config.py [hunk 1]
```diff
@@ -8,7 +8,10 @@ def migrate(config: dict):
         config["num_best_models_to_keep"] = config["num_models_to_keep"]
         del config["num_models_to_keep"]
 
-    return config
+    return {
+        "lr": 0.9,
+        **config,
+    }
```
Comment (Hiroshiba) [inline train/src/config.py:14]: 将来lrの値を別の値に代入するみたいになったとき、lrが値としてほしくなりそう。
意外とこうしといたほうが良いかもしれない･･･？

```suggestion
    if "lr" not in conifg:
        config["lr"] = 0.9

    return config
```

File: train/src/train.py [hunk 2]
```diff
@@ -269,6 +275,19 @@ def train():
...
+        total = 0
+        with torch.no_grad():
+            for eng, kata, e_mask, k_mask in tqdm(eval_dl, desc=f"Epoch {epoch} eval"):
+                out = model(eng, kata, e_mask, k_mask)
+                loss = criterion(out.transpose(1, 2), kata[:, 1:])
+                total_loss += loss.item()
+                total += 1
+
+        writer.add_scalar("Loss/eval", total_loss / total, epoch)
+        print(f"Epoch {epoch} Loss: {total_loss / total}")
```
Comment (Hiroshiba) [inline train/src/train.py:289]: これ実は計算が想定とずれる罠があります･･･！

例えばバッチサイズが64、データ数が65だったとき、
最初のバッチはデータが64個あるので1データ辺りのlossは1/64されてるけど、
次のバッチはデータが1個しかないので1/1なんですね。
つまり最後のデータほど指標への影響が大きくなってしまう。

より正確にするにはこういう感じにしないといけない。
```py
total_loss += loss.item() * len(out)
total += len(out)
```

ちなみに同じ理由で、学習時は最後のバッチだけ影響が大きくなってしまって不安定だから、DataLoaderで`drop_last=True`にしとくと良いかも。（忘れてた）
僕はよくこうしてます。

https://github.com/Hiroshiba/yukarin_sosfd/blob/6a29a9c35d8aa25b5e3092087963d9c29770b57c/train.py#L35-L50

File: train/src/train.py [hunk 3]
```diff
@@ -250,10 +250,16 @@ def train():
...
+        eval_dataset,
+        batch_size=batch_size,
+        shuffle=False,
+        collate_fn=partial(collate_fn, device=device),
+    )
 
     criterion = nn.CrossEntropyLoss(ignore_index=0)
     optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
-    scheduler = ExponentialLR(optimizer, 0.8)
+    scheduler = ExponentialLR(optimizer, config.lr)
```
Comment (Hiroshiba) [inline train/src/train.py:263]: ははーーーん！！　なんか違和感あったとおもったらこういうことか･･･！！

一般的にlrは上の行のOptimizer（今回はAdam）のlrを指します！
なのでこっち･･･！　`lr=1e-3`

`ExponentialLR`はcallされるたびにoptimizerのlrを定数倍するやつです。
https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ExponentialLR.html
> Decays the learning rate of each parameter group by gamma every epoch.

configの案としては･･･難しいですが、とりあえずこうとか？
```
exponential_lr_scheduler_gamma: 0.9
```

（ちなみに）
スケジューラーと呼ばれるものはだいぶいろいろあったりします。
最近だとこういうのも出てきてたり。
https://zenn.dev/dena/articles/6f04641801b387

Comment (sevenc-nanashi) [reply train/src/train.py]: exponential_lr_scheduler_gammaにしました。上のoptimizer_lrも設定できるようにしています。

File: train/config/example.yml [hunk 4]
```diff
@@ -16,3 +16,7 @@ num_best_models_to_keep: 2
 # 乱数のシード。
 seed: 0
 
+# Optimizerの学習率
+optimizer_lr: 0.001
+# 学習率の減衰率
+exponential_lr_scheduler_gamma: 0.9
```
Comment (Hiroshiba) [inline train/config/example.yml:22]: （ただのコメントです）

分けてる理由わからないですが、僕の感覚だとこれは上に含めちゃってて問題ないくらいの粒度かもです。

File: train/src/train.py (hunk 3 above)
Comment (Hiroshiba) [reply train/src/train.py]: optimizer_lrも設定できるの良いと思います！

**Review by Hiroshiba — APPROVED** [review]
LGTM！！

**Comment by sevenc-nanashi** [comment]
マージします。
